### PR TITLE
Charts caching 

### PR DESCRIPTION
**Pull request by EnoRage**
Implement caching for charts provider that allows to cache all charts data from start_time + duration.

There is an also case, when start_time is = 100 and the next request will be 99
So we clean cache if start time is lower than cached start time

The main idea for implementation: it does not change any behavior, even if we will have errors with caching - handler will work fine

P.S it is not perfect solution and we can cache data with worker later, however it will take much more time to implement and it also will take a lot of time to refactor current workers. According to the situation that this issue is critical - I think that current implementation is optimal for us

### CODE REVIEW

File: api/marketdata.go [hunk 1]
```diff
@@ -156,22 +155,61 @@ func getChartsHandler(charts *market.Charts) func(c *gin.Context) {
...
 		currency := c.DefaultQuery("currency", watchmarket.DefaultCurrency)
+
+		var isValidTimestamp bool
+		var response watchmarket.ChartData
+
+		key := cache.GenerateKey(coinQuery + token + maxItemsRaw + currency)
+
+		cacheData, err := cache.DB.Get(key)
+		if err != nil {
+			logger.Error("Failed to Get Cache", logger.Params{"err": err.Error()})
```
Comment (prazd) [inline api/marketdata.go]: Maybe return?

File: api/api.go [hunk 2]
```diff
@@ -5,11 +5,12 @@ import (
...
+	"github.com/trustwallet/watchmarket/services/caching"
 	"github.com/trustwallet/watchmarket/storage"
 )
 
-func Bootstrap(engine *gin.Engine, market storage.Market, charts *market.Charts, ac assets.AssetClient) {
+func Bootstrap(engine *gin.Engine, market storage.Market, charts *market.Charts, ac assets.AssetClient, cache *caching.Provider) {
 	engine.GET("/", func(c *gin.Context) { ginutils.RenderSuccess(c, `Watchmarket API`) })
 	marketAPI := engine.Group("/v1/market")
-	SetupMarketAPI(marketAPI, market, charts, ac)
+	SetupMarketAPI(marketAPI, market, charts, ac, cache)
```
Comment (arein) [inline api/api.go]: `cache` is an overloaded term here

File: api/api_test.go [hunk 3]
```diff
@@ -44,7 +45,7 @@ func TestTickers(t *testing.T) {
 	db := internal.InitRedis(fmt.Sprintf("redis://%s", s.Addr()))
 	seedDb(t, db)
 
-	Bootstrap(engine, db, getChartsMock(), getAssetClientMock())
+	Bootstrap(engine, db, getChartsMock(), getAssetClientMock(), nil)
```
Comment (arein) [inline api/api_test.go]: why special case the test?

File: api/marketdata.go [hunk 4]
```diff
@@ -1,28 +1,28 @@
...
 	"net/http"
 	"strconv"
 	"strings"
-	"time"
 )
 
 const (
 	defaultMaxChartItems = 64
-	tokenTWT             = "TWT-8C2"
+	minute               = 60
```
Comment (arein) [inline api/marketdata.go]: non-blocking: I think this is redundant - put the actual value here e.g. `cacheMaxAgeMinutes` or sth but not the definition of minute in seconds

File: api/marketdata.go [hunk 5]
```diff
@@ -156,12 +156,40 @@ func getChartsHandler(charts *market.Charts) func(c *gin.Context) {
...
+				return
+			}
+		}
+
+		if err == nil {
+			deleteResult, err := cache.DB.Delete(key)
+			if err != nil || deleteResult != storage.SaveResultSuccess {
+				logger.Warn("Failed to Delete Cache", logger.Params{"err": err})
+			}
+		}
```
Comment (arein) [inline api/marketdata.go]: this is too much logic for the controller. please encapsulate elsewhere

Comment (arein) [reply api/marketdata.go]: critical: you didn't test this in tests.

File: api/marketdata.go [hunk 6]
```diff
@@ -172,6 +200,17 @@ func getChartsHandler(charts *market.Charts) func(c *gin.Context) {
...
+		if err == nil && !chart.IsEmpty() {
+			saveResult, err := cache.DB.Set(key, storage.CacheData{
+				RawData:      rawData,
+				WasSavedTime: timeStart,
+			})
+			if err != nil || saveResult != storage.SaveResultSuccess {
+				logger.Error("failed to save cache", logger.Params{"key": key, "coin": coinId, "currency": currency, "token": token, "time_start": timeStart})
+			}
+		}
+
```
Comment (arein) [inline api/marketdata.go]: not tested also arguably shouldn't be in the controller but encapsulated elsewhere

File: tests/integration/setup/redis.go [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+			return errors.New("failed to init cache")
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func stopRedisContainer() error {
```
Comment (arein) [inline tests/integration/setup/redis.go]: the app should have no knowledge of a container concept. please rename.

File: tests/integration/setup/redis.go [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+		Cache = internal.InitRedis(fmt.Sprintf("redis://localhost:%s", redisResource.GetPort("6379/tcp")))
+		if Cache == nil {
+			return errors.New("failed to init cache")
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+	return nil
+}
```
Comment (arein) [inline tests/integration/setup/redis.go]: what is this solving? how is this related to adding cache? Should this be a dedicated PR? If so please move it into a dedicated PR?

this is also not tested?

**Comment by arein** [comment]
This PR lacks separation of concerns and testing. The controller does too much. There seem to be changes irrelevant to caching e.g. Redis retrying in this PR as well. Please be kind to your reviewers and split this out. This is also not motivated by the PR details.

File: api/api_test.go (hunk 3 above)
Comment (EnoRage) [reply api/api_test.go]: What do you mean ?

File: api/marketdata.go (hunk 5 above)
Comment (EnoRage) [reply api/marketdata.go]: There are tests (integration, also I cannot see any possibility to encapsulate it, it is tightly linked to the controller logic)

File: api/marketdata.go (hunk 6 above)
Comment (EnoRage) [reply api/marketdata.go]: Is tested and tightly linked to the controller logic

File: tests/integration/setup/redis.go (hunk 7 above)
Comment (EnoRage) [reply tests/integration/setup/redis.go]: It is integration test.

File: tests/integration/setup/redis.go (hunk 8 above)
Comment (EnoRage) [reply tests/integration/setup/redis.go]: It’s integration test

File: api/api.go (hunk 2 above)
Comment (EnoRage) [reply api/api.go]: what is more suitable option?

File: api/api_test.go (hunk 3 above)
Comment (EnoRage) [reply api/api_test.go]: We just reduce the time to init cache, we do not need it if we test another endpoint
